### PR TITLE
36 create navigation system

### LINKED ISSUE #36: Create Navigation System
Create a navigation folder that includes a navigation file aswell as a screen file.  Screen should include routing, so instead of manually using strings we use somethinge like this: Screen.MainScreen.Route, which essentially would be the route to mainpage. 
Navigation file should include a NavGraph (most likely NavHost will be used). This should include all forms of composable navigations throughout the application. This makes inclusion of new navigations easier, and makes the navigation set up more clean and readable for future implementations.

### PR DESCRIPTION
**Pull request by Steffenriv**
Created a new navigation system that centralizes all navigation in one folder system. Removes need for navhosts in other parts of the system. Updated files to make this system work for the whole architecture.
Closes #36

### CODE REVIEW

**Comment by Steffenriv** [comment]
Ignore duplicate commits, happened because i double rebased with local branch instead of origin and dev. My bad 👎

File: app/src/main/kotlin/com/platform/openemoji/navigation/Navigation.kt [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+                    arguments =
+                        listOf(
+                            navArgument("emojiTitle") {
+                                type = NavType.StringType
+                                nullable = true
+                            },
+                        ),
+                ) { backStackEntry ->
+                    val emojiTitle = backStackEntry.arguments?.getString("emojiTitle")
+                    val emoji = emojiTitle?.let { it1 -> emojiCatalogue.emoji(it1) }
```
Comment (OleAndreasER) [inline app/src/main/kotlin/com/platform/openemoji/navigation/Navigation.kt]: If you're going to name the parameter, give it a real name. "title"? The "it" from outside should also be named to "paddingValues".

File: app/src/main/kotlin/com/platform/openemoji/screens/SearchScreen.kt [hunk 2]
```diff
@@ -47,42 +43,26 @@ fun SearchScreen() {
...
+                    if (newCategory == overview) {
+                        emojiCatalogue.byCategory
+                    } else {
+                        emojiCatalogue.bySubCategory(newCategory) ?: emptyMap()
                     }
-                }
             }
+            EmojiCatalogueUi(
+                emojis = categorizedEmojis.value,
+                maxEmojisPerGrid = 15,
```
Comment (OleAndreasER) [inline app/src/main/kotlin/com/platform/openemoji/screens/SearchScreen.kt]: 15 emojis max is only supposed to be in "Overview".

File: app/src/main/kotlin/com/platform/openemoji/screens/SearchScreen.kt [hunk 3]
```diff
@@ -47,42 +43,26 @@ fun SearchScreen() {
...
-                emojiCatalogue.categories.forEach { category ->
-                    composable(
-                        "search/${route(category)}",
-                    ) {
-                        emojiCatalogue.bySubCategory(category)?.let {
-                            EmojiCatalogueUi(
-                                emojis = it,
-                            )
-                        }
+                listOf(stringResource(R.string.overview)) + emojiCatalogue.categories,
```
Comment (OleAndreasER) [inline app/src/main/kotlin/com/platform/openemoji/screens/SearchScreen.kt]: stringResource(R.string.overview) -> overview

File: app/src/main/kotlin/com/platform/openemoji/screens/SearchScreen.kt [hunk 4]
```diff
@@ -47,42 +43,26 @@ fun SearchScreen() {
                     color = MaterialTheme.colorScheme.onBackground,
                 )
                 val filteredEmojis = emojiCatalogue.search(query.value)
-                EmojiGrid(emojis = filteredEmojis)
+                EmojiGrid(emojis = filteredEmojis, navController = navController)
             }
         } else {
             // Remember the string resource here
```
Comment (OleAndreasER) [inline app/src/main/kotlin/com/platform/openemoji/screens/SearchScreen.kt]: This comment (not by you) is misleading and unnecessary. There is no remembering going on here.

File: app/src/main/kotlin/com/platform/openemoji/screens/SearchScreen.kt (hunk 2 above)
Comment (OleAndreasER) [reply app/src/main/kotlin/com/platform/openemoji/screens/SearchScreen.kt]: If you switch to remembering, not the categorizedEmojis, but just the category name, you can use that to determine max emojis.

File: app/src/main/kotlin/com/platform/openemoji/screens/SearchScreen.kt [hunk 5]
```diff
@@ -12,28 +12,24 @@ import androidx.compose.runtime.remember
...
 @Composable
-fun SearchScreen() {
+fun SearchScreen(navController: NavController) {
     val emojiCatalogue = EmojiCatalogue.get()
     // Creates a state for the search text
     val query = remember { mutableStateOf("") }
+    val categorizedEmojis = remember { mutableStateOf(emojiCatalogue.byCategory) }
 
     Column {
         SearchField(query)
```
Comment (OleAndreasER) [inline app/src/main/kotlin/com/platform/openemoji/screens/SearchScreen.kt:32]: This is not yours, but I think SearchField shouldn't have access to the mutable state "query", but instead should be provided a setter function: "SearchField { newQuery -> query.value = newQuery }".   

https://developer.android.com/jetpack/compose/architecture#udf

**Review by OleAndreasER — CHANGES_REQUESTED** [review]
Well done :)
- Returning from an emoji page should not reset the search page.  
- Bottom nav bar has different animation when pressing the text vs the icon.  
- Tests fail.  
- "NavigationItem" is not so descriptive. Isn't it a "Screen"? Or a "Route" (then also: Route.route -> Route.name)?

File: app/src/main/kotlin/com/platform/openemoji/screens/SearchScreen.kt (hunk 4 above)
Comment (Steffenriv) [reply app/src/main/kotlin/com/platform/openemoji/screens/SearchScreen.kt]: Removed it, does not need to be here, probably was added during rebase and i missed it when going over and did some minor tweaks.

File: app/src/main/kotlin/com/platform/openemoji/screens/SearchScreen.kt (hunk 3 above)
Comment (Steffenriv) [reply app/src/main/kotlin/com/platform/openemoji/screens/SearchScreen.kt]: Fixed!

File: app/src/main/kotlin/com/platform/openemoji/screens/SearchScreen.kt (hunk 2 above)
Comment (Steffenriv) [reply app/src/main/kotlin/com/platform/openemoji/screens/SearchScreen.kt]: Fixed, a bit wonky probably, but works well so.

File: app/src/main/kotlin/com/platform/openemoji/navigation/Navigation.kt (hunk 1 above)
Comment (Steffenriv) [reply app/src/main/kotlin/com/platform/openemoji/navigation/Navigation.kt]: Fixed!

File: app/src/main/kotlin/com/platform/openemoji/screens/SearchScreen.kt (hunk 5 above)
Comment (Steffenriv) [reply app/src/main/kotlin/com/platform/openemoji/screens/SearchScreen.kt:32]: fixed some minor issues now, will look at this a bit later alongside these points:
-Returning from an emoji page should not reset the search page.
-Bottom nav bar has different animation when pressing the text vs the icon.
Tests fail.
-"NavigationItem" is not so descriptive. Isn't it a "Screen"? Or a "Route" (then also: Route.route -> Route.name)? 
In regards to those points i have some questions:
-Quite interesting have not really thought about it. It does return to previous route (i.e. goes back to search page) used, so not really certain what i can do about it.
- No wonder lol, have put navigation it in the completly wrong place. (fixed)
-NavigationItem is actually quite common to use, but i agree it is not that good, can change to screen as i think that would suit best)

Comment (Steffenriv) [reply app/src/main/kotlin/com/platform/openemoji/screens/SearchScreen.kt:32]: In regards to test fail, which tests are you referring to?

Comment (OleAndreasER) [reply app/src/main/kotlin/com/platform/openemoji/screens/SearchScreen.kt:32]: EmojiCatalogueUiTest fails for me.

Comment (Steffenriv) [reply app/src/main/kotlin/com/platform/openemoji/screens/SearchScreen.kt:32]: Will look into that 👍

Comment (Steffenriv) [reply app/src/main/kotlin/com/platform/openemoji/screens/SearchScreen.kt:32]: @OleAndreasER  I have now fixed the test issue, as well Bottom bar different animation, swapped to screen from navigationItem. The only part where i am uncertain if i am able to actually fix properly is the refresh on return, i don't really know where to start.

File: app/src/main/kotlin/com/platform/openemoji/screens/SearchScreen.kt (hunk 2 above)
Comment (OleAndreasER) [reply app/src/main/kotlin/com/platform/openemoji/screens/SearchScreen.kt]: Here's what I imagined:  
```
    val overview = stringResource(R.string.overview)
    val category = remember { mutableStateOf(overview) }

```
```
            CategoryScrollCarousel(
                listOf(overview) + emojiCatalogue.categories,
            ) { newCategory ->
                category.value = newCategory
            }
```
```
            EmojiCatalogueUi(
                emojis =
                    if (category.value == overview) {
                        emojiCatalogue.byCategory
                    } else {
                        emojiCatalogue.bySubCategory(category.value) ?: emptyMap()
                    },
                maxEmojisPerGrid =
                    if (category.value == overview) {
                        15
                    } else {
                        null
                    },
                navController = navController,
            )

```

File: app/src/main/kotlin/com/platform/openemoji/screens/SearchScreen.kt (hunk 5 above)
Comment (OleAndreasER) [reply app/src/main/kotlin/com/platform/openemoji/screens/SearchScreen.kt:32]: Very good! Returning can be fixed in another issue. It shouldn't be slowing this issue down.

File: app/src/main/kotlin/com/platform/openemoji/screens/SearchScreen.kt (hunk 2 above)
Comment (OleAndreasER) [reply app/src/main/kotlin/com/platform/openemoji/screens/SearchScreen.kt]: Your solution works, so I will take care of this in another issue.

**Review by OleAndreasER — APPROVED** [review]

**Pull request merged**